src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -127,8 +128,9 @@ public IEnumerable<T> WalkWith<T>(PathSegmentTranslator<T> translator)
         [SuppressMessage("Microsoft.Design", "CA1011:ConsiderPassingBaseTypesAsParameters", Justification = "We would rather ship the PathSegmentHandler so that its more extensible later")]
         public void WalkWith(PathSegmentHandler handler)
         {
-            foreach (ODataPathSegment segment in this.segments)
+            for (var index = 0; index < this.segments.Count; index++)

```

Comment:
>this.segments.Count [](start = 40, length = 19)

perhaps move the computation of the count outside of the loop?
  int segments = this.segments.Count;
  for (int index=0; index < segments; index++)
   ...